test/sequence.test.js

```diff
@@ -0,0 +1,249 @@
+
+  assert.equal(tuples[1][2], 1, 'First array item should be 1');
+  assert.equal(tuples[2][2], 2, 'Second array item should be 2');
+
+  assert.equal(tuples[3][0], 7, 'Fourth tuple should be special type');
+  assert.equal(tuples[3][1], 31, 'Additional info should be 31 for break');
+});
+
+test('Nested structures in sequence', () => {
+  // Nested array [[[1]]] followed by 2, 3
+  // 81      -- Array (Length: 1)
+  //   81    --   [0] Array (Length: 1)
+  //     81  --     [0] Array (Length: 1)
+  //       01 --       [0] Unsigned: 1
+  // 02      -- Unsigned: 2
+  // 03      -- Unsigned: 3
+  const hex = '818181010203';
+  const seq = new Sequence(hexToU8(hex));
+
+  const tuples = Array.from(seq);
+  assert.equal(tuples.length, 6, 'Should have 6 tuples including nested structures');
+
+  // Check the nesting structure
+  assert.equal(tuples[0][0], 4, 'First tuple should be array type');
+  assert.equal(tuples[0][2], 1, 'Outer array length should be 1');
+
+  assert.equal(tuples[1][0], 4, 'Second tuple should be array type');
+  assert.equal(tuples[1][2], 1, 'Middle array length should be 1');
+
+  assert.equal(tuples[2][0], 4, 'Third tuple should be array type');
+  assert.equal(tuples[2][2], 1, 'Inner array length should be 1');
+
+  assert.equal(tuples[3][2], 1, 'Inner value should be 1');
+
+  // Then come the individual integers from the sequence
+  assert.equal(tuples[4][2], 2, 'Next sequence item should be 2');
+  assert.equal(tuples[5][2], 3, 'Last sequence item should be 3');
+});
+
+test('Tagged values in sequence', () => {
+  // Tag 0 (date/time string) with "2023-01-01", followed by 2, 3
+  // C0          -- Tag #0: (String Date)
+  //   6A        --   UTF8 (Length: 10): "2023-01-01"
+  //     323032332d30312d3031
+  // 02          -- Unsigned: 2
+  // 03          -- Unsigned: 3
+  const hex = 'c06a323032332d30312d30310203';
+  const seq = new Sequence(hexToU8(hex));
+
+  const tuples = Array.from(seq);
+
+  // First tuple should be tag 0
+  assert.equal(tuples[0][0], 6, 'First tuple should be tag type');
+  assert.equal(tuples[0][2], 0, 'Tag value should be 0 (date/time)');
+
+  // Second tuple should be the string
+  assert.equal(tuples[1][0], 3, 'Second tuple should be string type');
+  assert.equal(tuples[1][2], '2023-01-01', 'String value should be "2023-01-01"');
+
+  // The remaining sequence items
+  assert.equal(tuples[2][2], 2, 'Third item should be 2');
+  assert.equal(tuples[3][2], 3, 'Fourth item should be 3');
+});
+
+test('Incomplete CBOR data', () => {
+  // String claiming length 10 but only 5 bytes provided
+  // 6A          -- UTF8 (Length: 10)
+  //   48656c6c6f -- "Hello" (only 5 bytes)
+  const hex = '6a48656c6c6f';
+  const seq = new Sequence(hexToU8(hex));
+
+  // Should throw when trying to read past available bytes
+  assert.throws(() => seq.read(), {
+    message: /Unexpected end of stream/,
+  }, 'Should throw when reading incomplete string');
```

Comment:
Add another test for "8201", it gets a different error.

Reply:
`8201` meaning "array with 2 items"? That won't throw an error, instead we'll see "here comes an array with 2 items", "item one is `1`"… and the generator returns. The consumer needs to validate that it received the correct number of items.

Reply:
With this code:

```js
const s = new Sequence('8201');
[...s]
```

I get:
```
Uncaught RangeError: Offset is outside the bounds of the DataView
    at DataView.prototype.getUint8 (<anonymous>)
...
```

Reply:
Oh I see, it's because the test wasn't reading all items.

This is a problem, actually. `*#nextVal()` tries to read all array items so it throws when it's out of data. IMHO it'd make more sense if it only returned the next item. Same for maps and tags.

Reply:
Also the same for integers expecting 8 more bytes, that only get 6, for example.  If we could add more data, would good things happen?  At that point, we have most of a TransformStream<Uint8Array, any>.

Reply:
"Add more data" as in when it becomes available? I think we'll have already consumed the array header and preceding items, so we could resume from the last offset and take the next item.

At which point we're decoding one item at a time, since we'll no longer know we're in an array. So that should work.

May want a better error object though which includes offset details. Although it's also available in the last tuple.

If we don't change the behavior I think the `Sequence` documentation could do a better job at reflecting it. Ultimately "all data is expected to be present" would still hold, like it does for `decode()`.

Reply:
Let's just do the doc change for now, add a test or two for the current expected behavior, land this PR, and decide if we have a use case we care about for real streaming.